public/app/features/explore/Logs/LogsTableWrap.tsx

```diff
@@ -0,0 +1,278 @@
+import { css } from '@emotion/css';
+import { debounce } from 'lodash';
+import memoizeOne from 'memoize-one';
+import React, { useEffect } from 'react';
+
+import {
+  DataFrame,
+  ExploreLogsPanelState,
+  GrafanaTheme2,
+  Labels,
+  LogsSortOrder,
+  SplitOpen,
+  TimeRange,
+} from '@grafana/data/src';
+import { Themeable2 } from '@grafana/ui/src';
+
+import { fuzzySearch } from '../../../plugins/datasource/prometheus/querybuilder/components/metrics-modal/uFuzzy';
+import { parseLogsFrame } from '../../logs/logsFrame';
+
+import { LogsColumnSearch } from './LogsColumnSearch';
+import { LogsTable } from './LogsTable';
+import { LogsTableMultiSelect } from './LogsTableMultiSelect';
+
+interface Props extends Themeable2 {
+  logsFrames: DataFrame[];
+  width: number;
+  timeZone: string;
+  splitOpen: SplitOpen;
+  range: TimeRange;
+  logsSortOrder: LogsSortOrder;
+  panelState: ExploreLogsPanelState | undefined;
+  updatePanelState: (panelState: Partial<ExploreLogsPanelState>) => void;
+  onClickFilterLabel?: (key: string, value: string, refId?: string) => void;
+  onClickFilterOutLabel?: (key: string, value: string, refId?: string) => void;
+  datasourceType?: string;
+}
+
+export type fieldNameMeta = { percentOfLinesWithLabel: number; active: boolean | undefined };
+type fieldName = string;
+type fieldNameMetaStore = Record<fieldName, fieldNameMeta>;
+
+export const LogsTableWrap: React.FunctionComponent<Props> = (props) => {
+  const { logsFrames } = props;
+  // Save the normalized cardinality of each label
+  const [columnsWithMeta, setColumnsWithMeta] = React.useState<fieldNameMetaStore | undefined>(undefined);
+
+  // Filtered copy of columnsWithMeta that only includes matching results
+  const [filteredColumnsWithMeta, setFilteredColumnsWithMeta] = React.useState<fieldNameMetaStore | undefined>(
+    undefined
+  );
+
+  const dataFrame = logsFrames[0];
+
+  /**
+   * when the query results change, we need to update the columnsWithMeta state
+   * and reset any local search state
+   * @todo refactor
+   *
+   * This will also find all the unique labels, and calculate how many log lines have each label into the labelCardinality Map
+   * Then it normalizes the counts
+   *
+   */
+  useEffect(() => {
+    const numberOfLogLines = dataFrame ? dataFrame.length : 0;
```

Comment:
Same as the other component. If we can extract some of this complexity to a function or a few functions, I think it would be great.